juju/juju
# Charm origin

## Description of change

The following adds charm origin to state. This is independent of the
charmstore charm URL, hence why we want to own the origin for the
charm. 

Additionally, it adds the new changes but doesn't finish wiring them up.
Instead, it joins up the new charm origin, without parsing at every
stage. That will come later when we implement the download, refresh API.

Part of these changes includes the parsing and typing of a charm
channel. It conforms to the snapcraft channels docs[1] and collapses
where possible to be consistent with snap.

Both migration import and export is wired up, but again these are place
holders for when more data might be needed.

Additional work will be required:

 - Adding charm origin to show-application
 - Migrating state charm url and charm channel to charm origin.
   - Charm URL can be quite easily moved to charm origin ID.

1. https://snapcraft.io/docs/channels

## Issues:

 1. CharmURL **must** have a schema, we'll need to fix this in follow up
     PRs, as this is a right pain for upgrades and deployments.
 2. All charms **must** have a charm URL, failure to do so, will cause errors

## QA steps

```sh
juju bootstrap lxd test
juju deploy cs:~jameinel/ubuntu-lite-7
```

Access Mongo to see that the data is correctly stored.
